zaza/openstack/charm_tests/glance_simplestreams_sync/tests.py

```diff
@@ -125,3 +125,21 @@ def _check_local_product_streams(expected_images):
         _check_local_product_streams(expected_images)
 
         logging.debug("Local product stream successful")
+
+
+class GlanceSimpleStreamsSyncWithPropertiesTest(GlanceSimpleStreamsSyncTest):
+
+    @classmethod
+    def setUpClass(cls):
+        """Run class setup for running glance simple streams sync tests."""
+        super(GlanceSimpleStreamsSyncWithPropertiesTest, cls).setUpClass()
+        zaza_model.set_application_config('glance-simplestreams-sync', {'set_latest_property': 'true'})
```

Comment:
Why is this part of a class set up?  Wouldn't this line be better in the test.  Plus, I think you'll need to wait for the application to settle as this function is asynchronous.  i.e. it returns before the `config-changed` hook has completed (or even started) on the model. e.g. you'll need something like:

```python
    zaza_model.set_application_config(...)
    zaza_model.wait_for_application_states(states=states)
```

You'll need to work out what `states` is above for this particular test.  Take a look at other usages of `set_application_config` in test code (not in the utilities) to see what they are doing.

Reply:
I've use it in the test setup because I don't know how to tear down the images and rerun sync-images actions, or to force glance to sync the images on top of the existing one. 

if the images have been synced, already exist in glance and  we call sync-image action nothing will change as
 the `latest=true` property is attached to new images only. 

that's why I wrote a separated class for the test instead of a new test case. I also think the latter is better.

Reply:
Sorry, I wasn't clear: there's two issues here:

1. The setup of the property in the `setUpClass()` method; I don't think it should be there; it's actually part of the test rather than the fixture.
2. (More important): handling the async nature of `set_application_config()`;

Having the test as a separate class is fine.

Reply:
so, the reason why the setup of the property is in the `setUpClass()` is due the fact I need that config to be done BEFORE the image is actually sync. 
If we set the config after the images already exists in glance and call the sync-images action, nothing will be done. the existing image won't get the property.

Reply:
> so, the reason why the setup of the property is in the `setUpClass()` is due the fact I need that config to be done BEFORE the image is actually sync. If we set the config after the images already exists in glance and call the sync-images action, nothing will be done. the existing image won't get the property.

That doesn't make sense to me.  Image syncing is kicked off by `../setup.py:sync_images()`  How are you ensuring that the class is setup before that function is called?

Reply:
hmm. you're right. I'm not ensuring the config is set before the sync. actually the `setup.py:sync_images()` is called in the `functest-configure` phase.

so, should I 

1. set `set_latest_property=true` in the  `functest-configure` phase, and test if there is an image with latest property?
2. delete any existing image, `set_latest_property=true`, resync, and test if there is an image with latest property?
3. something else which I'm not seeing?

Thanks for pointing that out!

Reply:
I've done option 1. 

what you've said makes way more sense now :) 

thanks